### ISSUE TITLE
Communicate to user the status of the estimation returned by the minimization library (AdolC, betadiff, etc...)

### ISSUE THREAD

**Issue by mkienzle** [NIWAFisheriesModelling/CASAL2#233]
CASAL2, used in estimation mode, is designed to estimate the value of parameters using a minimization library (e.g. adolc, betadiff, etc...). Such libraries do not always converge to a satisfactory solution and the minimization algorithms implemented in mathematical libraries will tell you when they fail to do so: for example, a search might hit a boundary set by the user for a parameter; or the algorithm might use the maximum number of iterations it was allowed to use (using the max.iter setting) before it reaches the criteria of successful convergence; or it might not converge because the stopping criteria measuring the difference between successive values of the objective function (often refer to as the epsilon) does not satisfy the threshold setting. 

For the moment, CASAL2's users are not informed of the status of the estimation routine at the end of the minimization. Therefore CASAL2's users might very well be using results from a failed optimization.

It is necessary that CASAL2's users are provided with the return status of the minimization library in order for the users of this software to decide whether to use or discard results returned by CASAL2.

**Comment by smormede** [comment]
The casal2 manual is missing a portion of section 3 of casal manual about piping. Can @mkienzle or @SamikDatta please add to the manual (I assume it's the same). It's the bit in section3 starting with: CASAL uses both the standard output and standard error; we suggest redirecting both into files, but not if using CASAL in an interactive mode (for example, with -Y). Generally the two manuals need to be compared and a whole lot of stuff transferred to the new manual. 

basically run something like
casal2 -e > estimate.out > err.out
err.out will give you the warnings etc. @mkienzle can you try to see if it captures something in your simulations? I don't have a run with non successful convergence at hand.

However, if the convergence is successful in the command line there is something like
Successful convergence in optimise
@Zaita can this be captured somehow and added to one of the reports? 

thanks

**Comment by mkienzle** [comment]
thanks for looking into this @smormede 

a command such as you describe (**casal2 -e > estimate.out > err.out**) creates an empty estimate.out and redirect the stdout to the file err.out on my system (ubuntu linux). May be you want to recommend CASAL2's users to use something like **casal2 -e > estimate.out** instead.

As for the content of CASAL2 standard output relative to this issue (i.e. communicating to the user the status of the mathematical library after it finished the minimization), my system returns the following information:

...
*estimate_value[summary]
values {d}
process[Recruitment].r0 
2.60425e+06 
*end
...

which to me is *only* the value of the objective function.

Is anyone able to get the status of the minimization algorithm (in this case Adol-C) when it finishes its job? if so, can this someone, please, demonstrate how she/he does this?

**Comment by smormede** [comment]
@mkienzle check out the manual section 3.3 has the details for linux systems too, I believe:
(casal2 [arguments] > out) >& err &

We get if the convergence is successful or not in the cmd line. I have tested if the minimisation outputs are captured in the error file and they are not. Test is simple: reduce the number of iteration to say 10. It comes up with all sorts of messages on the command line that are not captured in err.out. 

@Zaita Can these be captured somewhere?

@mkienzle the standard output is for the minimised values only at this stage, plus if it has minimised as the message. If you want something else, please write a scoping document. It might be a standard output in the new minimiser you want to add to the libraries and an easy add. I understand from previous conversations that it is very minimiser-specific what the output might be.

**Comment by Zaita** [comment]
casal2 -e > estimate.out > err.out

This is saying "run casal2 with the -e parameter, then pipe stdout to estimate.out, actually pipe stdout to err.out instead.

You want to use
casal2 -e > estimate.out 2> err.out

**Comment by smormede** [comment]
@mkienzle just tested the above and I get in the file all the minimisation steps as well as if the minimiser converged or not. If that was not clear in the manual feel free to add a bit of text to make it clearer. 

Thanks @Zaita.

**Comment by mkienzle** [comment]
Hi all,

still don't get any more information with  casal2_adolc -e > estimate.out
2> err.out on my system.

Are you using Adol-C?
May be it's easier if I come to @smormede office tomorrow for a
demonstration

**Comment by mkienzle** [comment]
Adol-C in CASAL2 does not return the status of the minimiser at the end of the minisation: @Zaita is it because the Adol-C library does not provide the information? or because CASAL2 does not transfer the information from the library to the user?

Betadiff does provide users whether it converged or not at the end of its computation. If users need this information, it is recommended that they use Betadiff over Adol-C.

**Comment by Zaita** [comment]
I'm implementing a new "Estimation_Result" report. This will convey information where possible; the biggest issue is that it might not be 100% accurate (e.g. displaying convergence when it's a step size too small convergence) because every minimiser works differently and has different metrics for success/failure.

**Comment by Zaita** [comment]
Added.
